typed_python/PySetInstance.cpp

```diff
@@ -93,6 +100,288 @@ PyObject* PySetInstance::setClear(PyObject* o, PyObject* args) {
     Py_RETURN_NONE;
 }
 
+void PySetInstance::copy_elements(PyObject* dst, PyObject* src) {
+    PySetInstance* dst_w = (PySetInstance*)dst;
+
+    PyObjectStealer iterator(PyObject_GetIter(src));
+    if (!iterator) {
+        throw PythonExceptionSet();
+    }
+
+    PyObject* item;
+    while ((item = PyObjectStealer(PyIter_Next(iterator)))) {
+
+        if (!item) {
+            if (PyErr_Occurred()) {
+                throw PythonExceptionSet();
+            } else {
+                // no more values to iterate over
+                break;
+            }
+        }
+
+        Instance key(dst_w->type()->keyType(), [&](instance_ptr data) {
+            PyInstance::copyConstructFromPythonInstance(dst_w->type()->keyType(), data, item, true);
+        });
+
+        if (try_insert_key(dst_w, item, key.data()) != 0) {
+            throw PythonExceptionSet();
+        }
+    }
+}
+
+PyObject* PySetInstance::setCopy(PyObject* o, PyObject* args) {
+    if (args && PyTuple_Size(args) != 0) {
+        PyErr_SetString(PyExc_TypeError, "Set.copy takes no arguments");
+        return NULL;
+    }
+
+    PySetInstance* new_inst = (PySetInstance*)PyInstance::tp_new(Py_TYPE(o), NULL, NULL);
+    if (!new_inst) {
+        return NULL;
+    }
+    new_inst->mIteratorOffset = 0;
+    new_inst->mIteratorFlag = 0;
+    new_inst->mIsMatcher = false;
+
+    try {
+        copy_elements((PyObject*)new_inst, (PyObject*)(PySetInstance*)o);
+    } catch (PythonExceptionSet& e) {
+        decref((PyObject*)new_inst);
+        return NULL;
+    } catch (std::exception& e) {
+        decref((PyObject*)new_inst);
+        PyErr_SetString(PyExc_TypeError, e.what());
+        return NULL;
+    }
+    return (PyObject*)new_inst;
+}
+
+PyObject* PySetInstance::set_intersection(PyObject* o, PyObject* other) {
+    PySetInstance* new_inst = (PySetInstance*)PyInstance::tp_new(Py_TYPE(o), NULL, NULL);
+    if (!new_inst) {
+        return NULL;
+    }
+    new_inst->mIteratorOffset = 0;
+    new_inst->mIteratorFlag = 0;
+    new_inst->mIsMatcher = false;
+
+    PyObjectStealer iterator(PyObject_GetIter(other));
+    if (!iterator) {
+        decref((PyObject*)new_inst);
+        return NULL;
+    }
+
+    PySetInstance* self_w = (PySetInstance*)o;
+    PyObject* item;
+    while ((item = PyObjectStealer(PyIter_Next(iterator))) != NULL) {
```

Comment:
same comment as above, on both stealers and on needing a fastpath for other typed-python objects of appropriate type.

Reply:
done